File: week2/all_tasks.py
def find_and_print(messages): 
    # write down your judgment rules in comments 

    # if any keywords in their sentence, it would be considered as that the user is older than 17 years old
    # create a list of these keywords!

    # 'college student': general case is that all above 18 years old, but corner case would be genius student!

    # to be accurate, we should use the following keywords, but it is complicated for coding.
    # keywords = ['18 years old', 'legal age in Taiwan', 'vote']

    # to simplify the situation, we just use a single word for keywords
    keywords = ['18', 'legal', 'vote']
    
    # your code here, based on your own rules 

    for item in messages:

        
        words = messages[item].split()
        for word in words:
            if word in keywords:
                print(item)
                break
      
find_and_print({ 
"Bob":"My name is Bob. I'm 18 years old.", 
"Mary":"Hello, glad to meet you.", 
"Copper":"I'm a college student. Nice to meet you.", 
"Leslie":"I am of legal age in Taiwan.", 
"Vivian":"I will vote for Donald Trump next week", 
"Jenny":"Good morning." 
}) 


def calculate_sum_of_bonus(data): 
    # write down your bonus rules in comments 
    # Bonus is based on 3 things, salary, performance, and role fields.

    # Salary: base. just the numbers
    # performance. 3 levels. above/average/below.  the weighing factor: 30%/10%/20%
    # role fields: 3 levels. Individual contributors/managers/C-level. weighing factor: 10%/20%/30% 
    
    # your code here, based on your own rules 
    sum_bonus = 0

    dict_performance = {"above average": 0.03,
                        "average": 0.02,
                        "below average": 0.01,
                        }
    
    dict_role = {"CEO": "c_level",
                #  "Manager": "mid_level", # for future reference
                 "Engineer": "individual_contributor",
                 "Sales": "individual_contributor",
                 }
    
    dict_role_to_number = {"c_level": 0.03,
                           "mid_level": 0.02,
                           "individual_contributor": 0.01,
    }

    for item in data:
        for index in range(len(data[item])):
            if not isinstance(data[item][index]['salary'], int):
                data_salary = data[item][index]['salary']
                base = convert_to_integer(data_salary)

            else:
                base = data[item][index]['salary']

            bonus =  base * dict_performance[data[item][index]['performance']] + \
                        base * dict_role_to_number[dict_role[data[item][index]['role']]]

            sum_bonus += bonus
    
    print(sum_bonus)


def convert_to_integer(string_value):
    # Remove non-numeric characters and convert to integer
    numeric_value = int(''.join(filter(str.isdigit, string_value)))

    # Apply currency conversion if "USD" is present in the string
    if "USD" in string_value:
        numeric_value *= 30

    return numeric_value




calculate_sum_of_bonus({ 
"employees":[ 
    { 
    "name":"John", 
    "salary":"1000USD", 
    "performance":"above average", 
    "role":"Engineer" 
    }, 
    { 
    "name":"Bob", 
    "salary":60000, 
    "performance":"average", 
    "role":"CEO" 
    }, 
    { 
    "name":"Jenny", 
    "salary":"50,000", 
    "performance":"below average", 
    "role":"Sales" 
    } 
] 
})  # call calculate_sum_of_bonus function 


def func(*data): 
    # your code here
    # dictionary to document the frequency of names
    dict_middle_name = {}
    answer = []

    # Part 1: to create a lookup dictionary
    for name in data:
        if len(name) == 3 or len(name) == 2:
            dict_middle_name[name[1]] = dict_middle_name.get(name[1], 0) + 1
        else:
            print("Name is not right!")
            
    # Part 2: to print out the name whose frequency is only once
    # Part 2 is to be optimized. No need to go through all items again... Not sure how to simplify this part
    target_frequency = 1
    
    for name in data:
        if dict_middle_name[name[1]] == target_frequency:
            answer.append(name)
    
    if len(answer) == 0:
        print("沒有")
    else:
        for name in answer:
            print(name)



func("彭⼤牆", "王明雅", "吳明")  # print 彭⼤牆 
func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花")  # print 林花  花 
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花")  # print 沒有

def get_number(index):
    # your code here

    # this method is not that good. Brute force.
    # Optimized solution is recursion. But, I cannot make it for recursion...
    answer_array = []
    for item in range(index + 1):
        answer = 0
        if item % 2 == 0: # even number!
            answer = 3 * (item // 2)
        elif item == 1:
            answer = 4
        else:
            combo_1st_index = item // 2
            combo_2nd_index = combo_1st_index + 1
            answer = answer_array[combo_1st_index] + answer_array[combo_2nd_index]
        answer_array.append(answer)
    print(answer_array[index])

  
get_number(1)  # print 4 
get_number(5)  # print 10 
get_number(10)  # print 15 

File: week2/all_tasks.js
function findAndPrint(messages){ 
    // write down your judgment rules in comments 
    const arr_keywords = ['18', 'legal', 'vote'];
    // your code here, based on your own rules
    
    const keys = Object.keys(messages);
    // const Object.entries(messages)



 


    for (let i = 0; i < keys.length; i++) {
        let arr_words = messages[keys[i]];

        for (let j = 0; j < arr_keywords.length; j++) {
            if (arr_words.includes(arr_keywords[j])) {
                console.log(keys[i])
            }
            if (arr_words.includes(arr_keywords[j])) { break; }
        }
    }
    } 

findAndPrint({ 
"Bob":"My name is Bob. I'm 18 years old.", 
"Mary":"Hello, glad to meet you.", 
"Copper":"I'm a college student. Nice to meet you.", 
"Leslie":"I am of legal age in Taiwan.", 
"Vivian":"I will vote for Donald Trump next week", 
"Jenny":"Good morning." 
}); 


function convert_to_integer(string) {
    let numeric_value = string.replace(/[^0-9]/g, '');
    numeric_value = parseInt(numeric_value, 10);
    if (string.includes("USD")) {
        numeric_value *= 30;
    };
    return numeric_value;
}


function calculateSumOfBonus(data){

    // write down your bonus rule in comments 
    const dict_performance = {"above average": 0.03,
    "average": 0.02,
    "below average": 0.01,
    }

    const dict_role = {"CEO": "c_level",
    //   "Manager": "mid_level",  for future reference
    "Engineer": "individual_contributor",
    "Sales": "individual_contributor",
    }

    const dict_role_to_number = {"c_level": 0.03,
        "mid_level": 0.02,
        "individual_contributor": 0.01,
    }
        
    // your code here, based on your own rules

    let sum_bonus = 0;
    
    const keys_1st_level = Object.keys(data);
    const keys_2nd_level = Object.keys(data[keys_1st_level[0]]);
    // console.log(keys_1st_level);
    // console.log(keys_2nd_level);

    for (let i = 0; i < keys_1st_level.length; i++) {
        for(let j = 0; j < keys_2nd_level.length; j++) {
            // console.log(data[keys_1st_level[0]][j]['salary']);
            
            let temp_salary = data[keys_1st_level[0]][j]['salary'];
            if (temp_salary.constructor === Number) {
                base = data[keys_1st_level[i]][j]['salary'];
            } else { 
                base = convert_to_integer(temp_salary);

            }

            let bonus = base * dict_performance[data[keys_1st_level[i]][j]['performance']] + base * dict_role_to_number[dict_role[data[keys_1st_level[i]][j]['role']]];
            
            sum_bonus += bonus;


        }
    }
    console.log(sum_bonus);

    
    } 




calculateSumOfBonus({ 
"employees":[ 
{ 
"name":"John", 
"salary":"1000USD", 
"performance":"above average", 
"role":"Engineer" 
}, 
{ 
"name":"Bob", 
"salary":60000, 
"performance":"average", 
"role":"CEO" 
}, 
{ 
"name":"Jenny", 
"salary":"50,000", 
"performance":"below average", 
"role":"Sales" 
} 
] 
});  // call calculateSumOfBonus function 


function func(...data){ 
    // your code here 
    let dict_middle_name = {};
    let answer = [];


    // Part 1: to create a lookup dictionary
    for ([index, value] of data.entries()) {
        if (value.length === 3 || value.length === 2) {
            dict_middle_name[value[1]] = (dict_middle_name[value[1]] || 0) + 1; 
        } else {
            console.log("Name part has an issue!")
        }
    }

    // # Part 2: to print out the name whose frequency is only once
    // # Part 2 is to be optimized. No need to go through all items again... Not sure how to simplify this part
    target_frequency = 1;

    for ([index, value] of data.entries()) {
        if (dict_middle_name[value[1]] === target_frequency) {
            answer.push(value)
        }
    }

    if (answer.length === 0) {
        console.log("沒有");
    } else {
        for ([index, value] of answer.entries()) {
            console.log(value);
        }
    }

    } 




func("彭⼤牆", "王明雅", "吳明");  // print 彭⼤牆 
func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花");  // print  林花花 
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花");  // print 沒有 

function getNumber(index) { 
    // your code here 
    let answer_array = [];
    
    for (let item = 0; item <= index; item++) {
        let answer = 0
        if (item % 2 === 0) {
            answer = 3 * (Math.floor(item / 2))
        } else if (item == 1){
            answer = 4

        } else {
            let combo_1st_index = Math.floor(item / 2);
            let combo_2nd_index = combo_1st_index + 1;
            answer = answer_array[combo_1st_index] + answer_array[combo_2nd_index];
        }
        answer_array.push(answer);
        
    } 
    // console.log(answer_array);
    console.log(answer_array[index]);
    } 


getNumber(1);  // print 4 
getNumber(5);  // print 10 
getNumber(10);  // print 15 